### ISSUE TITLE
Style/CaseLikeIf false positive

### ISSUE THREAD

**Issue by noraj** [rubocop-hq/rubocop#8891]
## Related issues

#8541, https://github.com/rubocop-hq/rubocop/issues/7736#issuecomment-634915216, and mostly #8508

## Expected behavior

No offense when there is a something else than a comparison with a string/int, etc.
This is the same as #8508 where there was a comparison with a class but here it's `is_a?` method rather than the equal operator.

## Actual behavior

False positive when using a method, eg. is_a? with a class.

```ruby
if arg_feeds[0].is_a?(String)
 #
elsif arg_feeds[0].is_a?(Array)
 #
else
 #
end
```

## Steps to reproduce the problem

```
$ git clone https://gitlab.com/noraj/nvd_api.git && cd nvd_api
$ gem install bundle
$ ./bin/nvd_feed_api_setup
$ rake install
$ rubocop
```

![image](https://user-images.githubusercontent.com/16578570/95896369-6d950f80-0d8c-11eb-9d62-4bb5cfa6a44e.png)

## RuboCop version

```
$ bundle exec rubocop -V
0.92.0 (using Parser 2.7.1.5, rubocop-ast 0.7.1, running on ruby 2.7.1 x86_64-linux)
```

**Comment by koic** [comment]
The `case` statement uses `===` so it does not look like a false positive to `is_a?`.
https://ruby-doc.org/core-2.7.0/Module.html#method-i-3D-3D-3D

**Comment by noraj** [comment]
is_a? with a class is not a static value but will dynamically return true or false, so there can  be several "when" with true or false.

**Comment by noraj** [comment]
`is_a?` always return true or false.

```ruby
irb(main):002:0> 1.is_a?(Numeric)
=> true
irb(main):003:0> 1.is_a?(Integer)
=> true
```

So if you have:

```ruby
case var
when is_a?(Numeric)
 # command
when is_a?(Integer)
  # command
end
```

And you pass `var = 1`, both `when` statement will in fact be `when true`, I don't know if both will be executed or if a case-witch always execute only one statement.

I don't exactly understand why but when I switch from if/elsif to switch-case it broke all my unit test when it was using `is_a?`.

**Comment by noraj** [comment]
Take an example like here https://gitlab.com/noraj/nvd_api/-/blob/master/lib/nvd_feed_api.rb#L94-100

and change

```ruby
if arg_feeds[0].is_a?(String)
elsif arg_feeds[0].is_a?(Array)
```

into

```ruby
case arg_feeds[0]
when is_a?(String)
whenis_a?(Array)
```

And run `bundle exec rake install` before and after and you'll see before all is good and after nothing work.

**Comment by koic** [comment]
When auto-correction (`rubocop --only Style/CaseLikeIf -A`) is executed, it will be as follows.

```diff
diff --git a/lib/nvd_feed_api.rb b/lib/nvd_feed_api.rb
index 0c145f5..addbe2e 100644
--- a/lib/nvd_feed_api.rb
+++ b/lib/nvd_feed_api.rb
@@ -91,12 +91,13 @@ class NVDFeedScraper
     if arg_feeds.empty?
       return_value = @feeds
     elsif arg_feeds.length == 1
-      if arg_feeds[0].is_a?(String)
+      case arg_feeds[0]
+      when String
         @feeds.each do |feed| # feed is an object
           return_value = feed if arg_feeds.include?(feed.name)
         end
         # if nothing found return nil
-      elsif arg_feeds[0].is_a?(Array)
+      when Array
         raise 'one of the provided arguments is not a String' unless arg_feeds[0].all? { |x| x.is_a?(String) }

         # Sorting CVE can allow us to parse quicker
@@ -165,7 +166,8 @@ class NVDFeedScraper
     raise 'no argument provided, 1 or more expected' if arg_cve.empty?

     if arg_cve.length == 1
-      if arg_cve[0].is_a?(String)
+      case arg_cve[0]
+      when String
         raise 'bad CVE name' unless /^CVE-[0-9]{4}-[0-9]{4,}$/i.match?(arg_cve[0])

         year = /^CVE-([0-9]{4})-[0-9]{4,}$/i.match(arg_cve[0]).captures[0]
@@ -186,7 +188,7 @@ class NVDFeedScraper
         f = feeds(matched_feed)
         f.json_pull
         return_value = f.cve(arg_cve[0])
-      elsif arg_cve[0].is_a?(Array)
+      when Array
(snip)
```

So, the expected syntax is:

```ruby
case var
when Numeric
 # command
when Integer
  # command
end
```

How about it?
(Maybe it can be improved to the warning message.)

**Comment by fatkodima** [comment]
Correction:
The autocorrected code will be
```ruby
case var
when Numeric
 # command
when Integer
  # command
end
```

`case-when` executes the body of the first (and only one) truthy condition, like with `if-elsif`.
So, this is not a false positive.

**Comment by noraj** [comment]
So the idea is that `arg_feeds[0].is_a?(Array) == true` is equivalent to `arg_feeds[0].class == Array` so we can write `when String` in both cases. I applied with `rubocop --only Style/CaseLikeIf -A` and it doesn't break my code, thanks for the clarification.

**Comment by marcandre** [comment]
> So the idea is that `arg_feeds[0].is_a?(Array) == true` is equivalent to `arg_feeds[0].class == Array`

is equivalent to `Array === arg_feeds[0]`, and `when Array` is equivalent to `Array ===`
